File: 3121-count-the-number-of-special-characters-ii/solution.py
class Solution:
    def numberOfSpecialChars(self, word: str) -> int:
        # Track the last occurrence of lowercase letters and first occurrence of uppercase letters
        last_lower = [-1] * 26
        first_upper = [-1] * 26
        
        for i, char in enumerate(word):
            if char.islower():
                last_lower[ord(char) - ord('a')] = i
            else:
                idx = ord(char) - ord('A')
                # We only record the first occurrence of the uppercase letter
                if first_upper[idx] == -1:
                    first_upper[idx] = i
                    
        special_count = 0
        # Check conditions for each of the 26 English letters
        for i in range(26):
            if last_lower[i] != -1 and first_upper[i] != -1 and last_lower[i] < first_upper[i]:
                special_count += 1
                
        return special_count

File: 3121-count-the-number-of-special-characters-ii/solution.cpp
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
    int numberOfSpecialChars(string word) {
        // Track the last occurrence of lowercase letters and first occurrence of uppercase letters
        vector<int> last_lower(26, -1);
        vector<int> first_upper(26, -1);
        
        int n = word.length();
        for (int i = 0; i < n; ++i) {
            char c = word[i];
            if (c >= 'a' && c <= 'z') {
                last_lower[c - 'a'] = i;
            } else if (c >= 'A' && c <= 'Z') {
                // We only record the first occurrence of the uppercase letter
                if (first_upper[c - 'A'] == -1) {
                    first_upper[c - 'A'] = i;
                }
            }
        }
        
        int special_count = 0;
        // Check conditions for each of the 26 English letters
        for (int i = 0; i < 26; ++i) {
            if (last_lower[i] != -1 && first_upper[i] != -1 && last_lower[i] < first_upper[i]) {
                special_count++;
            }
        }
        
        return special_count;
    }
};